tools/slang-capability-generator/capability-generator-main.cpp

```diff
@@ -263,7 +540,71 @@ struct CapabilityDisjunction
 {
     List<CapabilityConjunction> conjunctions;
 
-    void addConjunction(const CapabilityConjunction& c)
+    void addConjunction(DiagnosticSink* sink, SourceLoc sourceLoc, CapabilitySharedContext& context, CapabilityConjunction& c)
+    {
+        if (c.isImpossible())
+            return;
+        bool cImpliesThis = false;
+        for (Index i = 0; i < conjunctions.getCount();)
+        {
+            // implied sets will be replaced
+            if (c.implies(conjunctions[i]))
+            {
+                cImpliesThis = true;
+                conjunctions.fastRemoveAt(i);
+            }
+            else
+                i++;
+        }
+        if (cImpliesThis)
+        {
+            conjunctions.add(_Move(c));
+            return;
+        }
+
+        for (Index i = 0; i < conjunctions.getCount();)
+        {
+            if (conjunctions[i].implies(c))
+            {
+                // subset is implied, we do not need to add it.
+                return;
+            }
+            else
+            {
+                // validate we are not creating a disjunction of same targets
+                if (conjunctions[i].shareTargetAndStageAtom(c, context))
+                {
+                    if (sink)
+                    {
+                        sink->diagnose(sourceLoc, Diagnostics::unionWithSameKeyAtomButNotSubset, conjunctions[i].toString(), c.toString());
+                        sink = nullptr;
+                    }
+                }
+                i++;
+            }
+        }
+        conjunctions.add(_Move(c));
+    }
+    void removeImplied()
+    {
+        for (Index i = 0; i < conjunctions.getCount(); i++)
+        {
+            for (Index ii = 0; ii < conjunctions.getCount(); ii++)
+            {
+                if (ii == i)
+                    continue;
+                if (conjunctions[i].implies(conjunctions[ii]))
+                {
+                    conjunctions.fastRemoveAt(ii);
```

Comment:
fastRemoveAt swaps element at ii with the last element. Could this lead to wrong result?

Reply:
You are correct, a incorrect result is possible.

```
i=getCount()-1;
ii=0
```

Would cause issues, regular 'remove' should be used.
* we remove as many times as overlapping sets, so normally only once, this should be fine.

1. `if [LAST_INDEX] => [0]`, we swap and delete 0.
2. now, `if [LAST_INDEX] => [1]`, we have a problem since [LAST_INDEX ] will never compare with [1]
* requires specific order

Reply:
I don't understand your explanation, Ariel.

What do you mean by "if last index is 0, we swap and delete 0"?
If last index is 0, it means the size of the container is 1 and the for-loop wouldn't even start, because there is nothing to compare to when there is only one element.

"if last index is 1, we have a problem since last index will never compare with 1".
I don't understand what this means.
If last index is 1, it means there are only two elements.
You will need to compare the first element and the second, and that's it.
Why would you want to compare the last element with the last element?

In general, "fast remove" type of remove is same as the regular remove except it changes the order of items in an unpredictable way, which is unpreferable in most cases.
Other than that, it is same as the regular remove.
In other words, if "fast remove" causes a problem, you should expect the same problem with the regular remove.

Reply:
I think there is a misunderstanding in what I mean by [LAST_INDEX]

Reply:
[LAST_INDEX] => [1]
means
conjunctions[LAST_INDEX].implies(conjunctions[1])

I was using `=>` as mathematical notation `implies`

Reply:
This is a reoccurring communication problem with you, Ariel.
You assume a certain expression by yourself and start using it without explaining what it means.
Please use a plain English or define your magic symbols so that people can read your formula.

Reply:
To define the implies symbol, the symbol is defined as per the mathematical relational symbol ["implication"](https://mathvault.ca/hub/higher-math/math-symbols/logic-symbols/)

Reply:
How about `[` and `]`?
Is there a mathematical use of them that I don't know of?

Reply:
`[]` was to represent indexing since the context of the problem is `if (conjunctions[i].implies(conjunctions[ii]))`

Reply:
the entire context of this file is "relational operations with code"